### PR TITLE
document socket.fdopen

### PR DESCRIPTION
**Pull request by vcunat**
I consider it quite an important function for e.g. systemd-based services, so I'd feel better if it were officially blessed (in docs).

I only spent a few minutes looking at the sources, trying to document the exact meaning, so it might not be accurate.  Still, it worked OK [for us](https://knot-resolver.readthedocs.io/en/latest/modules.html#dns-over-http-doh) in combination with `http.server.new()`

### CODE REVIEW

File: doc/cqueues.tex [hunk 1]
```diff
@@ -434,6 +434,9 @@ \subsubsection{\routine{cqueues.COMMIT}}
 \subsubsection[\fn{socket.listen}]{\fn{socket.listen\{ $\ldots$ \}}}
 	Like \fn{socket.listen} with list arguments, but takes a table of named arguments. See also \fn{socket.connect\{\}}.
 
+\subsubsection[\fn{socket.fdopen}]{\fn{socket.fdopen(FD number)}}
+	Open an existing bound file-descriptor.  That is useful e.g. for listening on sockets passed from systemd.
```
Comment (daurnimator) [inline doc/cqueues.tex]: No need for it to be bound

File: doc/cqueues.tex [hunk 2]
```diff
@@ -434,6 +434,9 @@ \subsubsection{\routine{cqueues.COMMIT}}
 \subsubsection[\fn{socket.listen}]{\fn{socket.listen\{ $\ldots$ \}}}
 	Like \fn{socket.listen} with list arguments, but takes a table of named arguments. See also \fn{socket.connect\{\}}.
 
+\subsubsection[\fn{socket.fdopen}]{\fn{socket.fdopen(FD number)}}
```
Comment (daurnimator) [inline doc/cqueues.tex]: s/FD number/fd/

File: doc/cqueues.tex (hunk 1 above)
Comment (vcunat) [reply doc/cqueues.tex]: OK.  I knew I had only vague notion about the constrains of the function.

**Comment by vcunat** [comment]
@daurnimator: do I understand the code correctly that the only difference of `dup` is that it duplicates the FD and owns the copy (i.e. closes it when the cqueues socket closes, or something like that).

**Comment by daurnimator** [comment]
> do I understand the code correctly that the only difference of `dup` is that it duplicates the FD and owns the copy (i.e. closes it when the cqueues socket closes, or something like that).

Yes: `dup` makes a copy while `fdopen` takes ownership.

> BTW, assuming that's so, I can't see why prefer `.dup` in the `systemd.socket` case.

You don't. as I said above: "systemd fd storage is a correct usecase" (for `.fdopen`)

**Comment by vcunat** [comment]
I'm not sure about the exact formulations... is this OK?  Feel free to amend and push directly a version you like.

**Comment by vcunat** [comment]
Added.  I actually didn't realize that you don't seem to have write access to this repo :-)

BTW, I didn't document that you can pass a table with `.fd` to these functions.  _I_ can't see an advantage when there's only one parameter.

**Pull request merged**